project/src/backend/sdl/SDLApplication.cpp

```diff
@@ -923,6 +932,83 @@ namespace lime {
 	}
 
 
+	int SDLApplication::BatchUpdate (int numEvents) {
+
+		SDL_Event* mouseMoved = NULL;
+
+		if (!active) {
+			return -1;
+		}
+		
+		queueLength = numEvents;
+
+		if (queueLength > queueMaxLength) {
+
+			if (NULL != eventQueue) {
+				SDL_free (eventQueue);
+			}
+
+			eventQueue = reinterpret_cast<SDL_Event*> (SDL_malloc (queueLength * sizeof (SDL_Event)));
+			if (NULL == eventQueue) {
+				queueLength = 0;
+				queueMaxLength = 0;
+			} else {
+				queueMaxLength = queueLength;
+			}
+		}
+		
+		int numPending = GetPendingEvents (eventQueue, queueLength);
+
+		// only allow GC free objects while we're handling events
+		if (isGCBlocking) System::GCExitBlocking ();
+		isGCBlocking = false;
+		
+		int nextEvent;
+		for (nextEvent = 0; nextEvent < numPending; ++nextEvent) {
+
+			if (SDL_MOUSEMOTION == eventQueue[nextEvent].type) {
+				// we'll only handle the last motion event in the batch
+				mouseMoved = &eventQueue[nextEvent];
+			} else {
+				HandleEvent (&eventQueue[nextEvent]);
+				eventQueue[nextEvent].type = -1;
+				if (!active) {
+					return -1;
+				}
+			}
+		}
+
+		// handle the mouse motion event if there is one this batch
+		if (NULL != mouseMoved) {
+			HandleEvent (mouseMoved);
```

Comment:
As I said elsewhere, I'm not a fan of doing this out of order. It could get skipped if the above loop returns. Another approach would be to add a second loop above the first:

```cpp
for (nextEvent = 0; nextEvent < numPending; ++nextEvent) {
    if (eventQueue[nextEvent].type == SDL_MOUSEMOTION) {
        if (mouseMoved != NULL) mouseMoved->type = -1;
        mouseMoved = &eventQueue[nextEvent];
    }
}
```

This will set all but the final mouse moved event's type to -1. Then the main loop would skip over those events, and execute all >= 0 events in the right order.